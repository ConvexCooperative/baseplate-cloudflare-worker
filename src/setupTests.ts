import "isomorphic-fetch";
import { jest } from "@jest/globals";
import { MockInstance } from "jest-mock";
import { OrgSettings } from "./getOrgSettings";
import { merge } from "lodash-es";

let mocks = {};

global.FOUNDRY_ENV = "prod";
global.FOUNDRY_MFE_HOST = "https://example.com/";

beforeEach(() => {
  global.FOUNDRY_MFE_HOST = "https://cdn.single-spa-foundry.com/apps/";

  mocks = {};

  const mainKv: MockCloudflareKV = {
    get: jest.fn(),
    mockKv(newMocks: KvMocks) {
      merge(mocks, newMocks);
    },
  };

  (mainKv.get as MockInstance).mockImplementation(async (key: string) => {
    for (let k in mocks) {
      if (k === key) {
        return mocks[k];
      }
    }

    if (key.startsWith("org-settings-")) {
      const orgSettings: Partial<OrgSettings> = {
        orgExists: true,
        staticFiles: {
          microfrontendProxy: {
            environments: {
              __main__: "https://example.com/",
            },
          },
        },
      };

      return orgSettings;
    }

    return false;
  });

  global.MAIN_KV = mainKv;
  global.FOUNDRY_ENV = "prod";
  global.FOUNDRY_MFE_HOST = "https://example.com/";
});

export interface MockCloudflareKV {
  get<ReturnValue>(key: string): Promise<ReturnValue>;
  mockKv(mocks: KvMocks): void;
}

interface KvMocks {
  [key: string]: any;
}
