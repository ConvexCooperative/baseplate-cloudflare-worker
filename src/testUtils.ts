import { RequestLog } from "./logRequests";

export function sampleLog(): RequestLog {
  return {
    customerEnv: "prod",
    httpStatus: 200,
    requestPath: "/hi",
    timestamp: Date.now(),
    userAgent: "Chrome",
  };
}
