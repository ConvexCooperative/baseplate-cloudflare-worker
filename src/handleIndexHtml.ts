import { getOrgSettings } from "./getOrgSettings";
import { RequestLog } from "./logRequests";
import { EnvVars } from "./main";
import { internalErrorResponse, notFoundResponse } from "./responseUtils";
import indexHtmlTemplate from "./indexHtml.mustache";
import { corsHeaders } from "./cors";
import { baseplateVersion } from "./baseplateVersion";
import Mustache from "mustache";
import {
  HTMLTemplateParams,
  mergeDefaultHtmlParams,
} from "@baseplate-sdk/utils";
import { readImportMap, verifyImportMap } from "./handleImportMap";
import * as singleSpa from "single-spa";
import { constructApplications, constructRoutes } from "single-spa-layout";
import { isCustomDomain } from "./customDomains";
import { parseFragment } from "parse5";

// Renders an HTML file to be used as a single-spa root config
export async function handleIndexHtml(
  request: Request,
  params: HandleIndexHtmlParams,
  requestLog: RequestLog,
  env: EnvVars,
  orgKey?: string
): Promise<Response> {
  if (!orgKey) {
    console.error(
      `No orgKey passed to handleIndexHtml function. Returning HTTP 500.`
    );
    return internalErrorResponse(request);
  }

  const [orgSettings, templateParameters] = await Promise.all([
    getOrgSettings(orgKey, env),
    env.MAIN_KV.get(`html-file-${orgKey}-${params.htmlFileName}`, {
      type: "json",
    }),
  ]);

  if (!orgSettings.orgExists || !templateParameters) {
    return notFoundResponse(request, orgSettings);
  }

  const finalParams: FinalHTMLTemplateParams =
    mergeDefaultHtmlParams(templateParameters);

  const importMap = await readImportMap(
    {
      importMapName: finalParams.importMap.name,
      customerEnv: params.customerEnv,
    },
    env,
    orgKey
  );
  const importMapErrors = verifyImportMap(importMap);

  if (importMapErrors.length > 0) {
    console.error(
      `Import Map Invalid for org ${orgKey} and request URL ${request.url}!`
    );
    console.error(importMapErrors);
    return internalErrorResponse(request, orgSettings);
  }

  // Add derived properties used in MustacheJS template
  finalParams.importMap.isSystemJS = finalParams.importMap.type === "systemjs";

  switch (finalParams.pageInit.type) {
    case "module":
      finalParams.pageInit.isSingleSpa = false;
      finalParams.pageInit.isEntryModule = true;
      break;
    case "single-spa":
      finalParams.pageInit.isSingleSpa = true;
      finalParams.pageInit.isEntryModule = false;

      const parsedLayout = parseFragment(finalParams.pageInit.layoutTemplate);
      const routes = constructRoutes(
        parsedLayout.childNodes[0] as unknown as Element
      );
      const applications = constructApplications({
        routes,
        async loadApp({ name }) {
          return { async bootstrap() {}, async mount() {}, async unmount() {} };
        },
      });
      const registerApplication =
        singleSpa["registerApplication"] ??
        // @ts-ignore
        singleSpa.default["registerApplication"];
      const checkActivityFunctions =
        singleSpa["checkActivityFunctions"] ??
        // @ts-ignore
        singleSpa.default["checkActivityFunctions"];
      const unregisterApplication =
        singleSpa["unregisterApplication"] ??
        // @ts-ignore
        singleSpa.default["unregisterApplication"];

      applications.forEach(registerApplication);
      let activeApplicationNames: string[], singleSpaLocation: URL;
      const requestUrl = new URL(request.url);
      if (isCustomDomain(requestUrl.hostname)) {
        singleSpaLocation = requestUrl;
      } else {
        const path = new URLSearchParams(requestUrl.search).get("path") ?? "/";
        singleSpaLocation = new URL(path, "https://example.com");
      }
      activeApplicationNames = checkActivityFunctions(
        // URL and Location objects are similar enough for single-spa to work with either
        singleSpaLocation as unknown as Location
      );
      applications.forEach((application) => {
        unregisterApplication(application.name);
      });

      for (let activeAppName of activeApplicationNames) {
        finalParams.preloads.push({
          as:
            finalParams.importMap.type === "native"
              ? "modulepreload"
              : "script",
          importSpecifier: activeAppName,
        });
      }
      break;
    default:
      console.error(
        // @ts-ignore
        `handleIndexHtml: htmlTemplateParameters.pageInit.type '${finalParams.pageInit.type}' is not supported.`
      );
      return internalErrorResponse(request, orgSettings);
  }

  finalParams.importMap.json = JSON.stringify(importMap!, null, 2);
  finalParams.importMap.url = new URL(
    `./${finalParams.importMap.name}.importmap`,
    request.url
  ).href;

  // Preload JS files from the import map that are known to be needed
  // Since browsers don't support import map specifiers in preload <link> elements,
  // we tell the browser to preload the full URL for the module, as found in the
  // import map
  for (let preload of finalParams.preloads) {
    if (preload.importSpecifier) {
      if (importMap?.imports[preload.importSpecifier]) {
        preload.href = importMap.imports[preload.importSpecifier];
        preload.as =
          finalParams.importMap.type === "native" ? "modulepreload" : "script";
      } else {
        console.error(
          `import specifier '${preload.importSpecifier}' cannot be preloaded because it doesn't exist in the import map`
        );
        return internalErrorResponse(request, orgSettings);
      }
    }
  }

  const html = Mustache.render(indexHtmlTemplate, finalParams);

  return new Response(html, {
    status: 200,
    headers: {
      "content-type": "text/html",
      // TODO: make this configurable as part of org settings
      "cache-control": "public, max-age=3600",
      // TODO: make this configurable as part of org settings or html params
      "content-security-policy":
        "default-src 'self' https: localhost:*; script-src 'unsafe-inline' 'unsafe-eval' https: localhost:*; connect-src https: localhost:* ws://localhost:*; style-src 'unsafe-inline' https:; object-src 'none';",
      ...corsHeaders(request, orgSettings),
      ...baseplateVersion(),
    },
  });
}

export interface HandleIndexHtmlParams {
  customerEnv: string;
  htmlFileName: string;
}

type FinalHTMLTemplateParams = HTMLTemplateParams & {
  importMap: HTMLTemplateParams["importMap"] & {
    isSystemJS?: boolean;
    json?: string;
    url?: string;
  };
  pageInit: HTMLTemplateParams["pageInit"] & {
    isSingleSpa: boolean;
    isEntryModule: boolean;
  };
};
