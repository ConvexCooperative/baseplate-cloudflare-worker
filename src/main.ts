import {
  match,
  MatchFunction,
  MatchResult,
  pathToRegexp,
} from "path-to-regexp";
import { handleImportMap } from "./handleImportMap";
import {
  notFoundResponse,
  internalErrorResponse,
  isCustomDomain,
} from "./responseUtils";
import { handleApps } from "./handleApps";
import { handleOptions } from "./cors";
import { logRequest, RequestLog } from "./logRequests";
import { handleIndexHtml, HandleIndexHtmlParams } from "./handleIndexHtml";
import { CustomDomain, CustomDomainPurpose } from "@baseplate-sdk/utils";

const workerHandler: ExportedHandler<EnvVars> = {
  fetch: handleRequest,
};

export default workerHandler;

const routeHandlers: RouteHandlers = withOptionalOrgKey({
  "/:customerEnv/:importMapName.importmap": handleImportMap,
  "/:customerEnv/apps/:pathParts*": handleApps,
  "/:customerEnv/:htmlFileName.html": handleIndexHtml,
});

const routeMatchers = Object.entries(routeHandlers).map(([path, handler]) => [
  match(path),
  handler,
]);

function withOptionalOrgKey(routeHandlers: RouteHandlers) {
  const result = {};
  for (const routeHandler in routeHandlers) {
    result[routeHandler] = routeHandlers[routeHandler];
    result["/:orgKey" + routeHandler] = routeHandlers[routeHandler];
  }

  return result;
}

const allowedMethods = ["GET", "HEAD", "OPTIONS"];

const requiredEnvironmentVariables: string[] = [
  "BASEPLATE_ENV",
  "AWS_REGION",
  "AWS_ACCESS_KEY_ID",
  "AWS_SECRET_ACCESS_KEY",
  "TIMESTREAM_DATABASE",
  "TIMESTREAM_TABLE",
  "MAIN_KV",
];

const orgKeyRegex = pathToRegexp("/:orgKey/(.*)");

export async function handleRequest(
  request: Request,
  env: EnvVars,
  context: ExecutionContext
) {
  const missingEnvVars = requiredEnvironmentVariables.filter(
    (requiredVar) => !env[requiredVar]
  );

  if (missingEnvVars.length > 0) {
    console.error(
      `Missing environment variables '${missingEnvVars.join(", ")}'`
    );
    return internalErrorResponse(request);
  }

  const requestUrl = new URL(request.url);
  let customDomain: CustomDomain | undefined;
  let orgKey: string | undefined;
  if (isCustomDomain(requestUrl.hostname)) {
    const kvKey = `custom-domain-${requestUrl.hostname}`;
    customDomain =
      (await env.MAIN_KV.get<CustomDomain>(kvKey, {
        type: "json",
      })) ?? undefined;

    orgKey = customDomain?.orgKey;
  } else {
    const match = orgKeyRegex.exec(requestUrl.pathname);
    orgKey = match && match[1] ? match[1] : undefined;
  }

  if (request.method === "OPTIONS") {
    return handleOptions(request, env, orgKey);
  } else if (!allowedMethods.includes(request.method)) {
    return notFoundResponse(request);
  }

  let response: Response,
    requestLog: RequestLog = {
      customerEnv: "unknown",
      requestPath: requestUrl.pathname,
      timestamp: Date.now(),
      userAgent: request.headers.get("user-agent"),
      httpStatus: 0,
    };

  // Baseplate CDN supports hosting single-spa root configs on custom domains
  // Normal Baseplate CDN route matching doesn't apply when on such a domain
  if (customDomain?.purpose === CustomDomainPurpose.web_app) {
    if (!customDomain.customerEnv) {
      console.error(
        `custom domain '${requestUrl.hostname}' does not have a customerEnv property in KV Storage`
      );
      return internalErrorResponse(request, undefined);
    } else if (!customDomain.webAppHtmlFilename) {
      console.error(
        `custom domain '${requestUrl.hostname}' does not have a webAppHtmlFilename property in KV Storage`
      );
      return internalErrorResponse(request, undefined);
    } else {
      // When the custom domain acts as a web app, we always call the
      // handleIndexHtml handler
      const params: HandleIndexHtmlParams = {
        customerEnv: customDomain.customerEnv,
        htmlFileName: customDomain.webAppHtmlFilename!,
      };

      return handleIndexHtml(request, params, requestLog, env, orgKey);
    }
  } else {
    // Normal Baseplate CDN route handling
    let routeHandler: RouteHandler | undefined,
      matchResult: MatchResult | false = false;

    // Find which route handler to call for this request
    // We don't have express to do this automatically for us, but are
    // using path-to-regexp which is what express uses under the hood
    for (let routeMatcher of routeMatchers) {
      const [match, handler] = routeMatcher;
      matchResult = match(requestUrl.pathname);
      if (matchResult) {
        routeHandler = handler;
        break;
      }
    }

    if (routeHandler && matchResult) {
      const params = {
        customerEnv: "prod",
        orgKey,
        ...matchResult.params,
      };
      requestLog.customerEnv = params.customerEnv;
      requestLog.orgKey = params.orgKey;
      response = await routeHandler(request, params, requestLog, env, orgKey);
    } else {
      response = await notFoundResponse(request);
    }
  }

  requestLog.httpStatus = response.status;

  context.waitUntil(logRequest(requestLog, env));
  return response;
}

interface RouteHandlers {
  [path: string]: RouteHandler;
}

type RouteHandler = (
  request: Request,
  params: object,
  requestLog: RequestLog,
  env: EnvVars,
  orgKey?: string
) => Promise<Response>;

export interface EnvVars {
  BASEPLATE_ENV: "dev" | "test" | "prod";
  AWS_REGION: string;
  AWS_ACCESS_KEY_ID: string;
  AWS_SECRET_ACCESS_KEY: string;
  TIMESTREAM_DATABASE: string;
  TIMESTREAM_TABLE: string;
  MAIN_KV: KVNamespace;
}
