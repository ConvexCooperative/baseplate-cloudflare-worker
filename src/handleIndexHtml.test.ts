import { HTMLTemplateParams } from "@baseplate-sdk/utils";
import { handleIndexHtml, HandleIndexHtmlParams } from "./handleIndexHtml";
import { EnvVars } from "./main";
import { createTestEnv } from "./setupTests";
import { sampleLog } from "./testUtils";
import { RecursivePartial } from "@baseplate-sdk/utils/lib/utils";

describe(`handleIndexHtml`, () => {
  let env: EnvVars,
    orgKey: string | undefined,
    params: HandleIndexHtmlParams,
    layoutTemplate;

  beforeEach(() => {
    env = createTestEnv();
    orgKey = "convex";
    params = {
      customerEnv: "prod",
      htmlFileName: "main",
    };
    layoutTemplate = `
    <single-spa-router>
      <nav class="topnav">
        <application name="@organization/nav"></application>
      </nav>
    </single-spa-router>`.trim();
  });

  it(`fails when no orgKey is passed`, async () => {
    const request = new Request(
      `https://cdn.baseplate.cloud/${orgKey}/prod/index.html`
    );
    orgKey = undefined;
    const response = await handleIndexHtml(
      request,
      params,
      sampleLog(),
      env,
      orgKey
    );

    expect(response.status).toBe(500);
  });

  it(`returns 404 when organization doesn't exist`, async () => {
    env.MAIN_KV.mockKv({
      [`org-settings-${orgKey}`]: null,
      [`html-file-${orgKey}-${params.htmlFileName}`]: {},
    });
    const request = new Request(
      `https://cdn.baseplate.cloud/${orgKey}/prod/index.html`
    );
    const response = await handleIndexHtml(
      request,
      params,
      sampleLog(),
      env,
      orgKey
    );

    expect(response.status).toBe(404);
  });

  it(`returns 404 when html file doesn't exist`, async () => {
    env.MAIN_KV.mockKv({
      [`html-file-${orgKey}-${params.htmlFileName}`]: null,
    });
    const request = new Request(
      `https://cdn.baseplate.cloud/${orgKey}/prod/index.html`
    );
    const response = await handleIndexHtml(
      request,
      params,
      sampleLog(),
      env,
      orgKey
    );

    expect(response.status).toBe(404);
  });

  it(`returns 500 if import map is missing in KV storage`, async () => {
    env.MAIN_KV.mockKv({
      [`html-file-${orgKey}-${params.htmlFileName}`]: {
        importMap: { name: "test" },
      },
      [`import-map-${orgKey}-test`]: null,
    });

    const request = new Request(
      `https://cdn.baseplate.cloud/${orgKey}/prod/index.html`
    );
    const response = await handleIndexHtml(
      request,
      params,
      sampleLog(),
      env,
      orgKey
    );

    expect(response.status).toBe(500);
  });

  it(`returns 500 if import map in KV Storage is not really an import map`, async () => {
    env.MAIN_KV.mockKv({
      [`html-file-${orgKey}-${params.htmlFileName}`]: {
        importMap: { name: "test" },
      },
      [`import-map-${orgKey}-test`]: { invalidKey: {} },
    });

    const request = new Request(
      `https://cdn.baseplate.cloud/${orgKey}/prod/index.html`
    );
    const response = await handleIndexHtml(
      request,
      params,
      sampleLog(),
      env,
      orgKey
    );

    expect(response.status).toBe(500);
  });

  it(`default index.html matches snapshot`, async () => {
    env.MAIN_KV.mockKv({
      [`html-file-${orgKey}-${params.htmlFileName}`]: {
        importMap: { name: "test" },
      },
      [`import-map-${orgKey}-${params.customerEnv}-test`]: {
        imports: {},
        scopes: {},
      },
    });

    const request = new Request(
      `https://cdn.baseplate.cloud/${orgKey}/prod/index.html`
    );
    const response = await handleIndexHtml(
      request,
      params,
      sampleLog(),
      env,
      orgKey
    );

    expect(response.status).toBe(200);
    expect(response.headers.get("content-type")).toBe("text/html");
    expect(response.headers.get("cache-control")).toBe("public, max-age=3600");
    expect(response.headers.get("content-security-policy")).toMatchSnapshot();
    expect(await response.text()).toMatchSnapshot();
  });

  it(`can render a SystemJS + single-spa root config`, async () => {
    const templateParameters: RecursivePartial<HTMLTemplateParams> = {
      importMap: { name: "test" },
      pageInit: {
        type: "single-spa",
        layoutTemplate,
      },
    };

    env.MAIN_KV.mockKv({
      [`html-file-${orgKey}-${params.htmlFileName}`]: templateParameters,
      [`import-map-${orgKey}-${params.customerEnv}-test`]: {
        imports: {},
        scopes: {},
      },
    });

    const request = new Request(
      `https://cdn.baseplate.cloud/${orgKey}/prod/index.html`
    );
    const response = await handleIndexHtml(
      request,
      params,
      sampleLog(),
      env,
      orgKey
    );

    expect(response.status).toBe(200);
    expect(await response.text()).toMatchSnapshot();
  });

  it(`can render a native module + single-spa root config`, async () => {
    const templateParameters: RecursivePartial<HTMLTemplateParams> = {
      importMap: { name: "test", type: "systemjs" },
      pageInit: {
        type: "single-spa",
      },
    };

    env.MAIN_KV.mockKv({
      [`html-file-${orgKey}-${params.htmlFileName}`]: templateParameters,
      [`import-map-${orgKey}-${params.customerEnv}-test`]: {
        imports: {},
        scopes: {},
      },
    });

    const request = new Request(
      `https://cdn.baseplate.cloud/${orgKey}/prod/index.html`
    );
    const response = await handleIndexHtml(
      request,
      params,
      sampleLog(),
      env,
      orgKey
    );

    expect(response.status).toBe(200);
    expect(await response.text()).toMatchSnapshot();
  });

  it(`can render a systemjs module + entryModule root config`, async () => {
    const templateParameters: RecursivePartial<HTMLTemplateParams> = {
      importMap: { name: "test", type: "systemjs" },
      pageInit: {
        type: "module",
        entryModule: "@walmart/root-config",
      },
    };

    env.MAIN_KV.mockKv({
      [`html-file-${orgKey}-${params.htmlFileName}`]: templateParameters,
      [`import-map-${orgKey}-${params.customerEnv}-test`]: {
        imports: {},
        scopes: {},
      },
    });

    const request = new Request(
      `https://cdn.baseplate.cloud/${orgKey}/prod/index.html`
    );
    const response = await handleIndexHtml(
      request,
      params,
      sampleLog(),
      env,
      orgKey
    );

    expect(response.status).toBe(200);
    expect(await response.text()).toMatchSnapshot();
  });

  it(`can render a native module + entryModule root config`, async () => {
    const templateParameters: RecursivePartial<HTMLTemplateParams> = {
      importMap: { name: "test", type: "native" },
      pageInit: {
        type: "module",
        entryModule: "@walmart/root-config",
      },
    };

    env.MAIN_KV.mockKv({
      [`html-file-${orgKey}-${params.htmlFileName}`]: templateParameters,
      [`import-map-${orgKey}-${params.customerEnv}-test`]: {
        imports: {},
        scopes: {},
      },
    });

    const request = new Request(
      `https://cdn.baseplate.cloud/${orgKey}/prod/index.html`
    );
    const response = await handleIndexHtml(
      request,
      params,
      sampleLog(),
      env,
      orgKey
    );

    expect(response.status).toBe(200);
    expect(await response.text()).toMatchSnapshot();
  });
});
