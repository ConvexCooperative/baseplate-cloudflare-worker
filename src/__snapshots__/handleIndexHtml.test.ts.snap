// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handleIndexHtml can render a SystemJS + single-spa root config 1`] = `
"<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <meta http-equiv="X-UA-Compatible" content="ie=edge">
  <title>Web App</title>
  <!-- Import Map URL: https://cdn.baseplate.cloud/convex/prod/test.importmap  -->
  <script type="systemjs-importmap">
    {
  "imports": {
    "single-spa": "https://cdn.jsdelivr.net/npm/single-spa@6.0.0/lib/es2015/system/single-spa.dev.min.js",
    "single-spa-layout": "https://cdn.jsdelivr.net/npm/single-spa-layout@2.2.0/dist/system/single-spa-layout.min.js",
    "react": "https://cdn.jsdelivr.net/npm/react@18.2.0/umd/react.production.min.js",
    "react-dom": "https://cdn.jsdelivr.net/npm/react-dom@18.2.0/umd/react-dom.production.min.js"
  },
  "scopes": {}
}
  </script>
  <meta name="importmap-type" content="systemjs-importmap" />
  <script src="https://cdn.jsdelivr.net/npm/systemjs@6.14.2/dist/system.min.js"></script>
  <script src="https://cdn.jsdelivr.net/npm/systemjs@6.14.2/dist/extras/amd.min.js"></script>
</head>
<body>
  <noscript>
    You need to enable JavaScript to run this app.
  </noscript>
  <main></main>
    <template id="single-spa-layout">
      <single-spa-router>
      <route path="/settings">
        <application name="@walmart/settings"></application>
      </route>
    </single-spa-router>
    </template>
    <script>
    Promise
      .all([System.import("single-spa"), System.import("single-spa-layout")])
      .then(([singleSpa, singleSpaLayout]) => {
        const { registerApplication, start } = singleSpa;
        const { constructApplications, constructRoutes, constructLayoutEngine } = singleSpaLayout;

    const routes = constructRoutes(document.querySelector('#single-spa-layout'));
    const applications = constructApplications({
      routes,
      loadApp({ name }) {
        return System.import(name);
      },
    });
    const layoutEngine = constructLayoutEngine({ routes, applications });

    applications.forEach(registerApplication);
    start();
      })
    </script>
  <script defer src="https://cdn.jsdelivr.net/npm/import-map-overrides@3.1.1/dist/import-map-overrides.js"></script>
  <import-map-overrides-full show-when-local-storage="devtools" dev-libs></import-map-overrides-full>
</body>
</html>"
`;

exports[`handleIndexHtml can render a native module + entryModule root config 1`] = `
"<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <meta http-equiv="X-UA-Compatible" content="ie=edge">
  <title>Web App</title>
  <!-- Import Map URL: https://cdn.baseplate.cloud/convex/prod/test.importmap  -->
  <script type="importmap">
    {
  "imports": {},
  "scopes": {}
}
  </script>
</head>
<body>
  <noscript>
    You need to enable JavaScript to run this app.
  </noscript>
  <main></main>
    <script>
      import('@walmart/root-config');
    </script>
  <script defer src="https://cdn.jsdelivr.net/npm/import-map-overrides@3.1.1/dist/import-map-overrides.js"></script>
  <import-map-overrides-full show-when-local-storage="devtools" dev-libs></import-map-overrides-full>
</body>
</html>"
`;

exports[`handleIndexHtml can render a native module + single-spa root config 1`] = `
"<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <meta http-equiv="X-UA-Compatible" content="ie=edge">
  <title>Web App</title>
  <!-- Import Map URL: https://cdn.baseplate.cloud/convex/prod/test.importmap  -->
  <script type="systemjs-importmap">
    {
  "imports": {
    "single-spa": "https://cdn.jsdelivr.net/npm/single-spa@6.0.0/lib/es2015/system/single-spa.dev.min.js",
    "single-spa-layout": "https://cdn.jsdelivr.net/npm/single-spa-layout@2.2.0/dist/system/single-spa-layout.min.js",
    "react": "https://cdn.jsdelivr.net/npm/react@18.2.0/umd/react.production.min.js",
    "react-dom": "https://cdn.jsdelivr.net/npm/react-dom@18.2.0/umd/react-dom.production.min.js"
  },
  "scopes": {}
}
  </script>
  <meta name="importmap-type" content="systemjs-importmap" />
  <script src="https://cdn.jsdelivr.net/npm/systemjs@6.14.2/dist/system.min.js"></script>
  <script src="https://cdn.jsdelivr.net/npm/systemjs@6.14.2/dist/extras/amd.min.js"></script>
</head>
<body>
  <noscript>
    You need to enable JavaScript to run this app.
  </noscript>
  <main></main>
    <template id="single-spa-layout">
      <single-spa-router></single-spa-router>
    </template>
    <script>
    Promise
      .all([System.import("single-spa"), System.import("single-spa-layout")])
      .then(([singleSpa, singleSpaLayout]) => {
        const { registerApplication, start } = singleSpa;
        const { constructApplications, constructRoutes, constructLayoutEngine } = singleSpaLayout;

    const routes = constructRoutes(document.querySelector('#single-spa-layout'));
    const applications = constructApplications({
      routes,
      loadApp({ name }) {
        return System.import(name);
      },
    });
    const layoutEngine = constructLayoutEngine({ routes, applications });

    applications.forEach(registerApplication);
    start();
      })
    </script>
  <script defer src="https://cdn.jsdelivr.net/npm/import-map-overrides@3.1.1/dist/import-map-overrides.js"></script>
  <import-map-overrides-full show-when-local-storage="devtools" dev-libs></import-map-overrides-full>
</body>
</html>"
`;

exports[`handleIndexHtml can render a systemjs module + entryModule root config 1`] = `
"<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <meta http-equiv="X-UA-Compatible" content="ie=edge">
  <title>Web App</title>
  <!-- Import Map URL: https://cdn.baseplate.cloud/convex/prod/test.importmap  -->
  <script type="systemjs-importmap">
    {
  "imports": {},
  "scopes": {}
}
  </script>
  <meta name="importmap-type" content="systemjs-importmap" />
  <script src="https://cdn.jsdelivr.net/npm/systemjs@6.14.2/dist/system.min.js"></script>
  <script src="https://cdn.jsdelivr.net/npm/systemjs@6.14.2/dist/extras/amd.min.js"></script>
</head>
<body>
  <noscript>
    You need to enable JavaScript to run this app.
  </noscript>
  <main></main>
    <script>
      System.import('@walmart/root-config');
    </script>
  <script defer src="https://cdn.jsdelivr.net/npm/import-map-overrides@3.1.1/dist/import-map-overrides.js"></script>
  <import-map-overrides-full show-when-local-storage="devtools" dev-libs></import-map-overrides-full>
</body>
</html>"
`;

exports[`handleIndexHtml default index.html matches snapshot 1`] = `"default-src 'self' https: localhost:*; script-src 'unsafe-inline' 'unsafe-eval' https: localhost:*; connect-src https: localhost:* ws://localhost:*; style-src 'unsafe-inline' https:; object-src 'none';"`;

exports[`handleIndexHtml default index.html matches snapshot 2`] = `
"<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <meta http-equiv="X-UA-Compatible" content="ie=edge">
  <title>Web App</title>
  <!-- Import Map URL: https://cdn.baseplate.cloud/convex/prod/test.importmap  -->
  <script type="systemjs-importmap">
    {
  "imports": {
    "single-spa": "https://cdn.jsdelivr.net/npm/single-spa@6.0.0/lib/es2015/system/single-spa.dev.min.js",
    "single-spa-layout": "https://cdn.jsdelivr.net/npm/single-spa-layout@2.2.0/dist/system/single-spa-layout.min.js",
    "react": "https://cdn.jsdelivr.net/npm/react@18.2.0/umd/react.production.min.js",
    "react-dom": "https://cdn.jsdelivr.net/npm/react-dom@18.2.0/umd/react-dom.production.min.js"
  },
  "scopes": {}
}
  </script>
  <meta name="importmap-type" content="systemjs-importmap" />
  <script src="https://cdn.jsdelivr.net/npm/systemjs@6.14.2/dist/system.min.js"></script>
  <script src="https://cdn.jsdelivr.net/npm/systemjs@6.14.2/dist/extras/amd.min.js"></script>
</head>
<body>
  <noscript>
    You need to enable JavaScript to run this app.
  </noscript>
  <main></main>
    <template id="single-spa-layout">
      <single-spa-router></single-spa-router>
    </template>
    <script>
    Promise
      .all([System.import("single-spa"), System.import("single-spa-layout")])
      .then(([singleSpa, singleSpaLayout]) => {
        const { registerApplication, start } = singleSpa;
        const { constructApplications, constructRoutes, constructLayoutEngine } = singleSpaLayout;

    const routes = constructRoutes(document.querySelector('#single-spa-layout'));
    const applications = constructApplications({
      routes,
      loadApp({ name }) {
        return System.import(name);
      },
    });
    const layoutEngine = constructLayoutEngine({ routes, applications });

    applications.forEach(registerApplication);
    start();
      })
    </script>
  <script defer src="https://cdn.jsdelivr.net/npm/import-map-overrides@3.1.1/dist/import-map-overrides.js"></script>
  <import-map-overrides-full show-when-local-storage="devtools" dev-libs></import-map-overrides-full>
</body>
</html>"
`;
